### ISSUE TITLE
Not able to use flush policies with async logger

### ISSUE THREAD

**Issue by ravidhiman** [gabime/spdlog#2670]
Neither timeout based not log level based flushing working on below code

spdlog::init_thread_pool(mAcroLogConfig->mQueueSize, mAcroLogConfig->mWorkerThread );
            auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_st>(fileName.c_str(), 1048576 * mAcroLogConfig->mMaxFileSize, mAcroLogConfig->mRotatingFiles);
            file_sink->set_pattern("%^[%Y.%m.%d] [%H:%M:%S.%F] [%n:%l] [ProcessID:%P] [ThreadId:%t] %v%$");
            mLogger = std::make_shared<spdlog::async_logger>(ALM_DEFAULT_LOGGER_NAME, file_sink, spdlog::thread_pool(), spdlog::async_overflow_policy::overrun_oldest/*queue full policy*/);
            spdlog::register_logger(mLogger);
            mLogger->set_level(spdlog::level::level_enum(mAcroLogConfig->mLogLevel));

            if (mAcroLogConfig->mFlushPolicy == LogLevelBased){
                mLogger->flush_on((spdlog::level::level_enum)mAcroLogConfig->mFlushLogLevel);
            }
            else {
                spdlog::flush_every(std::chrono::seconds(mAcroLogConfig->mFlushTimeout));
            }

**Comment by ravidhiman** [comment]
@gabime

**Comment by tt4g** [comment]
What do you mean by "Not able to use flush policies with async logger"?
Please explain the expected and actual behavior.

Note that the async logger is also flushed asynchronously.

**Comment by ravidhiman** [comment]
@tt4g  Basically, if i set flush timer to 1 minutes. It doesnt work. It still flush continuously. Ideally it should flush after 1 minute
spdlog::flush_every(std::chrono::seconds(60))
and same for flush level. if i set flush level to error, it still flush for info and debug. Idealy it should flush only when error log comes
mLogger->flush_on((spdlog::level::level_enum)(4))

**Comment by ravidhiman** [comment]
Complete working code for log level flushing:

#include<iostream>
#include<windows.h>
#include <spdlog/spdlog.h>
#include <spdlog/sinks/stdout_color_sinks.h>
#include <spdlog/sinks/basic_file_sink.h>
#include <spdlog/sinks/rotating_file_sink.h>
#include <spdlog/async.h>
using namespace std;

int main()
{

    spdlog::init_thread_pool(1024,1 );
    auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_st>("abc", 1048576 * 1, 3);
    file_sink->set_pattern("%^[%Y.%m.%d] [%H:%M:%S.%F] [%n:%l] [ProcessID:%P] [ThreadId:%t] %v%$");
    auto mLogger = std::make_shared<spdlog::async_logger>("abc", file_sink, spdlog::thread_pool(), spdlog::async_overflow_policy::overrun_oldest/*queue full policy*/);
    spdlog::register_logger(mLogger);
    mLogger->set_level(spdlog::level::level_enum(SPDLOG_LEVEL_TRACE));    
    mLogger->flush_on(spdlog::level::level_enum(SPDLOG_LEVEL_ERROR));
    

    for (int i = 0; i < 100; i++)
        mLogger->debug("hi");
    mLogger->error("error");

    for (int i = 0; i < 1000000000; i++)
    {
        mLogger->info("info");
    }

    mLogger->error("error");


	return 0;
}

**Comment by tt4g** [comment]
`spdlog::logger::flush_on()` (or `spdlog::flush_on()`) is an API that sets the sink to trigger a flush when logging messages greater than or equal to a specified log level.
It does not filter out flushes at lower log levels.
And `spdlog::flush_every()` flushes sink periodically, so if you are calling the two APIs, both flush by log level and periodic flush will be triggered.

Furthermore, even if the sink is not flushed, the OS may automatically flush it before the data buffer for writing to the file is full.
In other words, while you can explicitly force flushing, you cannot freely control the timing of flushing (because the OS automatically performs the flushing).

**Comment by ravidhiman** [comment]
thanks @tt4g  Got the answer to spdlog::flush_on(). So means we are forcefully flushing this.
I didnt get "so if you are calling the two APIs, both flush by log level and periodic flush will be triggered" this part. You mean that for period flush do i need to call flush by log level and periodic flush both?
mLogger->flush_on((spdlog::level::level_enum)(4))
spdlog::flush_every(std::chrono::seconds(60))

**Comment by tt4g** [comment]
It means that sink is flushed by the next two triggers.

1. when logging above the error level (`mLogger->flush_on((spdlog::level::level_enum)(4))`)
2. every 60 seconds (`spdlog::flush_every(std::chrono::seconds(60))`)

**Comment by ravidhiman** [comment]
ok thanks.That clarifies. Also i have noticed that if logs are less, flush still doesnt show the logs in file. Is it expected?

**Comment by tt4g** [comment]
On Windows, we know that if the application exits before the buffer is flushed, any log data remaining in the buffer will be discarded.
To avoid this problem, call `spdlog::shutdown()` at the end of the `main()` function.

This problem does not occur on Linux OS and Mac OS.

**Comment by ravidhiman** [comment]
yes i am calling shutdown. But issue is when application crashes, then we kind of lose logs.

**Comment by tt4g** [comment]
Undefined behavior after crash due to C++ specification. It is not guaranteed to be flushed after a crash.

First, make sure that crashes do not occur.
If you cannot avoid crashes, register a signal handler to handle crash signals (#1607).
Nevertheless, the C++ specification does not guarantee that the application will work after a crash, even with signal handling.
